### PR TITLE
Fix oom during content update process

### LINKED ISSUE #22141: Linux Out of Memory Killer Triggered on Minimal Resource Wazuh Dedicated Manager Host
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| 4.8.0-beta2 | Vulnerability Detection | Manager | Packages (AIO) | Ubuntu 22 |

### Manager resources

| CPU |  Memory |
| --- | --- |
| 2  | 4098 | 

### Description

A significant spike in RAM consumption has been observed during the processing of vd_updater content. This surge has led to the failure of the dedicated manager host with minimal Indexer requirements due to a Linux Out of Memory killer.

```
root@ubuntu22-1:/home/vagrant# systemctl status wazuh-manager
× wazuh-manager.service - Wazuh manager
     Loaded: loaded (/lib/systemd/system/wazuh-manager.service; enabled; vendor preset: enabled)
     Active: failed (Result: oom-kill) since Mon 2024-02-26 15:09:39 UTC; 33min ago
    Process: 122370 ExecStart=/usr/bin/env /var/ossec/bin/wazuh-control start (code=exited, status=0/SUCCESS)
    Process: 124294 ExecStop=/usr/bin/env /var/ossec/bin/wazuh-control stop (code=exited, status=0/SUCCESS)
        CPU: 32min 43.530s

Feb 26 15:09:36 ubuntu22-1 env[124294]: Killing wazuh-db...
Feb 26 15:09:37 ubuntu22-1 env[124294]: Killing wazuh-authd...
Feb 26 15:09:38 ubuntu22-1 env[124294]: wazuh-agentlessd not running...
Feb 26 15:09:38 ubuntu22-1 env[124294]: wazuh-integratord not running...
Feb 26 15:09:38 ubuntu22-1 env[124294]: wazuh-dbd not running...
Feb 26 15:09:38 ubuntu22-1 env[124294]: wazuh-csyslogd not running...
Feb 26 15:09:38 ubuntu22-1 env[124294]: Killing wazuh-apid...
Feb 26 15:09:39 ubuntu22-1 env[124294]: Wazuh v4.8.0 Stopped
Feb 26 15:09:39 ubuntu22-1 systemd[1]: wazuh-manager.service: Failed with result 'oom-kill'.
Feb 26 15:09:39 ubuntu22-1 systemd[1]: wazuh-manager.service: Consumed 32min 43.530s CPU time.

-----

× wazuh-indexer.service - Wazuh-indexer
     Loaded: loaded (/lib/systemd/system/wazuh-indexer.service; enabled; vendor preset: enabled)
     Active: failed (Result: oom-kill) since Mon 2024-02-26 14:45:48 UTC; 1h 10min ago
       Docs: https://documentation.wazuh.com
    Process: 67286 ExecStart=/usr/share/wazuh-indexer/bin/systemd-entrypoint -p ${PID_DIR}/wazuh-indexer.pid --quiet (code=killed, signal=KILL)
   Main PID: 67286 (code=killed, signal=KILL)
        CPU: 6min 16.096s

Feb 26 14:03:35 ubuntu22-1 systemd-entrypoint[67286]: WARNING: System::setSecurityManager will be removed in a future release
Feb 26 14:03:36 ubuntu22-1 systemd-entrypoint[67286]: WARNING: A terminally deprecated method in java.lang.System has been called
Feb 26 14:03:36 ubuntu22-1 systemd-entrypoint[67286]: WARNING: System::setSecurityManager has been called by org.opensearch.bootstrap.Security (file:/usr/share/wazuh-indexer/lib/opensearch->
Feb 26 14:03:36 ubuntu22-1 systemd-entrypoint[67286]: WARNING: Please consider reporting this to the maintainers of org.opensearch.bootstrap.Security
Feb 26 14:03:36 ubuntu22-1 systemd-entrypoint[67286]: WARNING: System::setSecurityManager will be removed in a future release
Feb 26 14:03:54 ubuntu22-1 systemd[1]: Started Wazuh-indexer.
Feb 26 14:45:48 ubuntu22-1 systemd[1]: wazuh-indexer.service: A process of this unit has been killed by the OOM killer.
Feb 26 14:45:48 ubuntu22-1 systemd[1]: wazuh-indexer.service: Main process exited, code=killed, status=9/KILL
Feb 26 14:45:48 ubuntu22-1 systemd[1]: wazuh-indexer.service: Failed with result 'oom-kill'.
Feb 26 14:45:48 ubuntu22-1 systemd[1]: wazuh-indexer.service: Consumed 6min 16.096s CPU time.
```

This issue has also impacted Vulnerability End-to-End (E2E) tests, rendering one of the manager inaccessible during testing sessions.

### PR DESCRIPTION
**Pull request by Dwordcito**
|Related issue|
|---|
|Closes #22141|

## Description

This pr aims to solve some problems of high memory usage when the content is updated or generated.

### CODE REVIEW

File: src/shared_modules/utils/jsonArrayParser.hpp [hunk 1]
```diff
@@ -54,7 +54,7 @@ namespace JsonArray
          */
         JsonSaxArrayParser(
             nlohmann::json::json_pointer targetArrayPointer,
-            std::function<bool(nlohmann::json&&)> itemCallback,
+            std::function<bool(nlohmann::json&&, const size_t)> itemCallback,
```
Comment (GabrielEValenzuela) [inline src/shared_modules/utils/jsonArrayParser.hpp:57]: I think we can add a document reference about the change in the callback

**Review by GabrielEValenzuela — COMMENTED** [review]
Great job Octa! One little comment.

The logic is clear and easy to follow!

**Review by sebasfalcone — DISMISSED** [review]
- Code 🟡 
  - One little change requested about the use of magic numbers
  
- Test 🟢 
  - I've processed 23K offsets (in 1K chunks) the memory is released periodically.
  - Some 200MB spikes are detected, they are related to high volume data
  
  
image

**Review by sebasfalcone — APPROVED** [review]
LGTM!

**Pull request merged**